### ISSUE TITLE
Spurious confusing errors on trait impl on type name collision

### ISSUE THREAD

**Issue by Havvy** [rust-lang/rust#46579]
```rust
type Type = ();

#[derive(Debug)]
struct Type;

impl Iterator for Type {}
```

Gives:

```
   Compiling playground v0.0.1 (file:///playground)
error[E0428]: the name `Type` is defined multiple times
 --> src/main.rs:4:1
  |
1 | type Type = ();
  | --------------- previous definition of the type `Type` here
...
4 | struct Type;
  | ^^^^^^^^^^^^ `Type` redefined here
  |
  = note: `Type` must be defined only once in the type namespace of this module

error[E0119]: conflicting implementations of trait `std::fmt::Debug` for type `()`:
 --> src/main.rs:3:10
  |
3 | #[derive(Debug)]
  |          ^^^^^
  |
  = note: conflicting implementation in crate `core`

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/main.rs:6:1
  |
6 | impl Iterator for Type {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
  |
  = note: the impl does not reference any types defined in this crate
  = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/main.rs:3:10
  |
3 | #[derive(Debug)]
  |          ^^^^^ impl doesn't use types inside crate
  |
  = note: the impl does not reference any types defined in this crate
  = note: define and implement a trait or new type instead

error: aborting due to 4 previous errors

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

The first error is true. The rest are spurious. There are probably other spurious errors that can be generated via name collision.

**Comment by estebank** [comment]
Current output, barely a few changes:

```
error[E0428]: the name `Type` is defined multiple times
 --> src/lib.rs:4:1
  |
1 | type Type = ();
  | --------------- previous definition of the type `Type` here
...
4 | struct Type;
  | ^^^^^^^^^^^^ `Type` redefined here
  |
  = note: `Type` must be defined only once in the type namespace of this module

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/lib.rs:3:10
  |
3 | #[derive(Debug)]
  |          ^^^^^
4 | struct Type;
  |        ---- this is not defined in the current crate because tuples are always foreign
  |
  = note: impl doesn't have any local type before any uncovered type parameters
  = note: for more information see https://doc.rust-lang.org/reference/items/implementations.html#orphan-rules
  = note: define and implement a trait or new type instead

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
 --> src/lib.rs:6:1
  |
6 | impl Iterator for Type {}
  | ^^^^^^^^^^^^^^^^^^----
  |                   |
  |                   this is not defined in the current crate because tuples are always foreign
  |
  = note: impl doesn't have any local type before any uncovered type parameters
  = note: for more information see https://doc.rust-lang.org/reference/items/implementations.html#orphan-rules
  = note: define and implement a trait or new type instead
```
